### ISSUE TITLE
Feature Request: Manual/OnChange trigger for metadata.yaml template creation/rerendering

### ISSUE THREAD

**Issue by monkz** [lxc/lxd#8152]
# Problem description

As of LXD 4.8 there are only 3 options when to trigger a render for a template. [copy,start,create]
If the configuration of a container changes, while it is running no change is applied.
This results in inconsistencies if the template mechanismn is used for cloud-init.
(Like in the "ubuntu:" containers. A change is only applied there if the container is created or copied.)

Example:
A container is launched with user.network-config set to use IP Address X for configuration/installation purposes, and must not be restarted.
The (now configured) container is ready to switch into the production network (IP Address Y is needed there).
A change to user.network-config is made. But the template using this variable is not rerendered, thus rerunning the cloud-init seed with the old IP Address X instead of IP Address Y. Generating once more the netplan configuration for IP Address X.

# Feature Idea
Create 2 additional ways to trigger a rerender of templates:
## Manual
A command is needed to rerender the templates like
`lxc metadata templates render <container> [templatename]`
if templatename is empty, all templates are to be rerendered.

This can be accompanied with a flag `--dry-run` to only show the result of the template
or `lxc metadata templates show <container> [templatename]`

## OnChange
If a variable (referenced in a template) is changed via `lxc config edit <container>` or `lxc config set <container> <key>=<value>` the template will be rerendered.
This is maybe the consumer friendliest method, but also the one with the most complexity.

# TL;DR
Please create a way (or ways) to reflect config changes on a container to be reflected immediately inside the container.

**Comment by stgraber** [comment]
There is a way to manually re-run the templates, it's the `volatile.apply_template` key which is processed on the next instance start. Doing `lxc config set blah volatile.apply_template create` will re-run all templates that use the `create` trigger.

`on-change` would certainly be great to have, but we have no idea what variable a template may use. All LXD does is expose a bunch of context to the template, we don't know what's being pulled from that context.

For the specific case of network config, the plan is for the cloud-init team to implement a full LXD backend, deprecating those templates and config keys in favor of first class cloud-init integration in LXD.

Closing as those two options should address your current and future needs.

**Comment by monkz** [comment]
Thank you

**Comment by monkz** [comment]
Hi @stgraber , after a long time I need to reopen this issue albeit for different reasons.
I need to render/generate templates while the container is running.
Basically
```
lxc config template render [<remote>:]<instance> <template> [flags]
```

**Comment by stgraber** [comment]
@monkz sorry for the delay.

I've been thinking about this on and off for the past few days and I'm struggling to find a safe way to allow this as well as provide something which will work the same way across both containers and virtual machines. In general, the template mechanism was designed to render files in the instance prior to it starting with the various triggers letting you decide on what kind of startups it should kick in.

The fact that we know the instance is guaranteed to be off allowed us to run a traditional templating engine (pongo2) directly from LXD without fearing a hostile container trying to mess with us as we interact with its filesystem.

Changing this to be able to run in a hostile environment makes things significantly harder as we'd then need to move that logic to a dedicated subprocess (`forktemplate` or something along those lines), implement a C loader that performs the critical security bits and then chain to the Go code to run the actual template. Templates in such a confined environment would also be unable to perform includes or access any other files, potentially breaking some existing templates.

This is effectively having to change the templating logic to work within the same confined environment that `lxc file push` and `lxc file pull` use. One in which we can't trust symlinks, we can't trust mounts and we can't trust any past results as a hostile user in the instance may be actively messing with the filesystem under us.

Those are the implications for doing this for containers. On the VM front, things are a bit easier because we don't really need to trust the guest, we'd just tell the agent to re-run the template. This would need some added logic for us to re-generate/re-populate the config drive that the VM agent uses and adding an API to have the agent reload the template file and apply it.

An additional catch that both would need to handle is that the file being templated could be currently locked or actively being read/written to by another process. This may need additional logic to first record attributes of the existing file, delete the existing file, template the file as a new one and then reset the attributes. Well, some non-racy variation of this anyway.


So I guess the TLDR here is that it's not impossible to do, but it'd be a lot of security sensitive work that'd need to be done for a very edge case feature and effectively just saving a manual edit through `lxc file` or a quick instance restart.

**Comment by monkz** [comment]
Thank you for your extensive answer.  We'll resort to template outside of LXD and `lxc file push` it.
